### ISSUE TITLE
Segmentation fault on Windows with OCaml 4.12

### ISSUE THREAD

**Issue by yallop** [ocaml/ocaml#10477]
The [Ctypes library](https://github.com/ocamllabs/ocaml-ctypes/) has a [test](https://github.com/ocamllabs/ocaml-ctypes/blob/master/tests/test-lifetime/test_lifetime.ml) that checks that code generated by the library keeps objects alive while the runtime lock is released.  The test consistently passes with all combinations of OCaml versions and operating systems, except one: [OCaml 4.12 on Windows](https://github.com/ocamllabs/ocaml-ctypes/issues/677), where it fails with a segmentation fault.

Some investigation suggests that the bug is due to a problem in OCaml, introduced between 4.11 and 4.12.  The following cut-down program (also available [here](https://github.com/yallop/ocaml-windows-segfault)) doesn't use ctypes, but does trigger segmentation faults fairly reliably:

```ocaml
let alloc () =
  for i = 0 to 20000 do
    for i = 0 to 200; do ignore (Array.make 10 ()) done;
    ignore (Array.make 1000 ());
    if i mod 1000 = 0 then (Gc.compact ());
  done

let churn () =
  let allocators = List.init 10 (fun _ -> Thread.create alloc ()) in
  List.iter Thread.join allocators

let () = churn ()
```

As the CI logs for the program show, [on 4.11 the test runs reliably](https://github.com/yallop/ocaml-windows-segfault/runs/2915044196?check_suite_focus=true#step:5:21), while [on 4.12 it segfaults every few runs](https://github.com/yallop/ocaml-windows-segfault/runs/2915044243?check_suite_focus=true#step:5:25).

**Comment by nojb** [comment]
Could reproduce (with `mingw64`):

![image](https://user-images.githubusercontent.com/113560/123445414-237d5280-d5d8-11eb-887d-8911f8718d58.png)

This points to 

https://github.com/ocaml/ocaml/blob/71166020951ffc2c3b351ae9536e7249380aa509/otherlibs/systhreads/st_stubs.c#L526

Perhaps related to #10161 ?

**Comment by dra27** [comment]
At a glance on my phone, it looks like that call to `st_set_thread_id` is on the wrong side of `caml_leave_blocking_section` 😱

**Comment by nojb** [comment]
Fix in #10478
